src/3d/qgs3daxis.cpp

```diff
@@ -0,0 +1,644 @@
+  clearBuffers->setBuffers( Qt3DRender::QClearBuffers::DepthBuffer );
+
+  return twoDViewport;
+}
+
+QVector3D Qgs3DAxis::from3dTo2dLabelPosition( const QVector3D &sourcePos,
+    Qt3DRender::QCamera *sourceCamera, Qt3DRender::QViewport *sourceViewport,
+    Qt3DRender::QCamera *destCamera, Qt3DRender::QViewport *destViewport,
+    const QSize &destSize )
+{
+  QVector3D destPos = sourcePos.project( sourceCamera->viewMatrix(),
+                                         destCamera->projectionMatrix(),
+                                         QRect( 0.0f, 0.0f,
+                                             destViewport->normalizedRect().width() * destSize.width(),
+                                             destViewport->normalizedRect().height() * destSize.height() ) );
+  QPointF axisCenter = sourceViewport->normalizedRect().center();
+  QPointF labelCenter = destViewport->normalizedRect().center();
+  QVector3D viewTranslation = QVector3D( ( axisCenter - labelCenter ).x() * destSize.width(),
+                                         ( axisCenter - labelCenter ).y() * destSize.height(),
+                                         0.0 );
+  destPos -= QVector3D( labelCenter.x() * destSize.width(),
+                        labelCenter.y() * destSize.height(),
+                        0.0f );
+  destPos.setX( destPos.x() + viewTranslation.x() );
+  destPos.setY( destPos.y() - viewTranslation.y() );
+  destPos.setZ( 0.0f );
+
+#ifdef DEBUG
+  qDebug() << "sourcePos" << sourcePos << " with" << viewTranslation << "corrected destPos" << destPos;
+#endif
+  return destPos;
+}
+
+void Qgs3DAxis::setEnableCube( bool show )
+{
+  mCubeRoot->setEnabled( show );
+}
+
+void Qgs3DAxis::setEnableAxis( bool show )
+{
+  mAxisRoot->setEnabled( show );
+  mText_X->setEnabled( show );
+  mText_Y->setEnabled( show );
+  mText_Z->setEnabled( show );
+}
+
+void Qgs3DAxis::createAxisScene()
+{
+  if ( mAxisRoot == nullptr || mCubeRoot == nullptr )
+  {
+#ifdef DEBUG
+    qDebug() << "Should recreate mAxisRoot" << mMode;
+#endif
+    mAxisRoot = new Qt3DCore::QEntity;
+    mAxisRoot->setParent( mAxisSceneEntity );
+    mAxisRoot->setObjectName( "3DAxisRoot" );
+
+    createAxis( Axis::X );
+    createAxis( Axis::Y );
+    createAxis( Axis::Z );
+
+    mCubeRoot = new Qt3DCore::QEntity;
+    mCubeRoot->setParent( mAxisSceneEntity );
+    mCubeRoot->setObjectName( "3DCubeRoot" );
+
+    createCube( );
+  }
+
+  if ( mMode == Mode::Off )
+  {
+    mAxisSceneEntity->setEnabled( false );
+    setEnableAxis( false );
+    setEnableCube( false );
+  }
+  else
+  {
+    mAxisSceneEntity->setEnabled( true );
+    if ( mMode == Mode::Crs )
+    {
+      setEnableCube( false );
+      setEnableAxis( true );
+      if ( mCrs.isGeographic() )
+      {
+        mText_X->setText( QStringLiteral( "Long" ) );
+        mText_Y->setText( QStringLiteral( "Lat" ) );
+      }
+      else
+      {
+        mText_X->setText( "X" );
+        mText_Y->setText( "Y" );
```

Comment:
Here you should be using mCrs.axisOrdering() to determine the actual axis for the crs. You can use this alongside QgsCoordinateReferenceSystemUtils::axisDirectionToAbbreviatedString to get the correct E/N suffix for the axis.

If we are showing correctly E/N/ etc as the crs axis, do we still need the "NorthEastUp" mode? Couldn't we just always use the actual CRS axis labels? (Unless the crs is invalid, in which case x/y/z is appropriate)

Reply:
I did not know this function. It will be helpful! By the way, the "NorthEastUp" mode is useless.